exchanges/gateio/gateio_websocket.go

```diff
@@ -38,697 +67,904 @@ func (g *Gateio) WsConnect() error {
 	for i := range channelsToSubscribe {
-		// Ensures params are in order
-		params := []interface{}{channelsToSubscribe[i].Currency}
-		if strings.EqualFold(channelsToSubscribe[i].Channel, "depth.subscribe") {
+		var auth *WsAuthInput
+		timestamp := time.Now()
+		var params []string
+		switch channelsToSubscribe[i].Channel {
+		case optionsUnderlyingTickersChannel,
+			optionsUnderlyingTradesChannel,
+			optionsUnderlyingPriceChannel,
+			optionsUnderlyingCandlesticksChannel:
+			var uly string
+			uly, err = g.GetUnderlyingFromCurrencyPair(channelsToSubscribe[i].Currency)
+			if err != nil {
+				return nil, err
+			}
+			params = append(params, uly)
+		case optionsBalancesChannel:
+			// options.balance channel does not require neither underlying not contract
+		default:
+			channelsToSubscribe[i].Currency.Delimiter = currency.UnderscoreDelimiter
+			params = append(params, channelsToSubscribe[i].Currency.String())
+		}
+		switch channelsToSubscribe[i].Channel {
+		case spotOrderbookChannel:
+			interval, okay := channelsToSubscribe[i].Params["interval"].(kline.Interval)
+			if !okay {
+				return nil, errors.New("invalid interval parameter")
+			}
+			level, okay := channelsToSubscribe[i].Params["level"].(int)
+			if !okay {
+				return nil, errors.New("invalid spot order level")
+			}
+			intervalString, err = g.GetIntervalString(interval)
+			if err != nil {
+				return nil, err
+			}
 			params = append(params,
-				channelsToSubscribe[i].Params["limit"],
-				channelsToSubscribe[i].Params["interval"])
-		} else if strings.EqualFold(channelsToSubscribe[i].Channel, "kline.subscribe") {
-			params = append(params, channelsToSubscribe[i].Params["interval"])
-		}
-
-		for j := range payloads {
-			if payloads[j].Method == channelsToSubscribe[i].Channel {
-				switch {
-				case strings.EqualFold(channelsToSubscribe[i].Channel, "depth.subscribe"):
-					if len(payloads[j].Params) == 3 {
-						// If more than one currency pair we need to send as
-						// matrix
-						_, ok := payloads[j].Params[0].(currency.Pair)
-						if ok {
-							var bucket = payloads[j].Params
-							payloads[j].Params = nil
-							payloads[j].Params = append(payloads[j].Params, bucket)
-						}
-					}
-
-					payloads[j].Params = append(payloads[j].Params, params)
-				case strings.EqualFold(channelsToSubscribe[i].Channel, "kline.subscribe"):
-					// Can only subscribe one market at the same time, market
-					// list is not supported currently. For multiple
-					// subscriptions, only the last one takes effect.
-				default:
-					payloads[j].Params = append(payloads[j].Params, params...)
-				}
-				payloads[j].Channels = append(payloads[j].Channels, channelsToSubscribe[i])
-				continue channels
+				strconv.Itoa(level),
+				intervalString,
+			)
+		case futuresOrderbookChannel:
+			limit, ok := channelsToSubscribe[i].Params["limit"].(int)
+			if !ok {
+				return nil, errors.New("invlaid futures orderbook limit")
```

Comment:
```suggestion
				return nil, errors.New("invalid futures orderbook limit")
```
Could probably make it an actual error variable instead